File: web/src/types/twitter-downloader.ts
/** Interface for TweetMeta */
export interface TweetMeta {
    /** URL of the tweet */
    tweetUrl: string

    /** ID of the user who uploaded the tweet */
    uploaderUserId: string

    /** Display name of the user who uploaded the tweet */
    uploaderDisplayName: string

    /** Date the tweet was uploaded */
    uploadDate: string

    /** Duration of the tweet media */
    duration: string

    /** URL of the thumbnail for the tweet media */
    thumbnail: string

    /** Title of the tweet */
    title: string

    /** Description of the tweet */
    description: string
}

/** Interface for TweetMedia */
export interface TweetMedia {
    /** Format of the tweet media */
    format: string

    /** Width of the tweet media */
    width: number

    /** Height of the tweet media */
    height: number

    /** Resolution of the tweet media */
    resolution: string

    /** URL of the tweet media */
    url: string

    /** Size of the tweet media */
    size: string
}

/** Interface for TwitterDownloaderResult */
export interface TwitterDownloaderResult {
    /** Meta data for the tweet */
    tweetMetaData: TweetMeta

    /** Array of media information for the tweet */
    tweetMedias: TweetMedia[]

    /** Error message, if any */
    error: string
}

/** Interface for TwitterDownloaderResponse */
export interface TwitterDownloaderResponse {
    /** Result of the Twitter downloader operation */
    result: TwitterDownloaderResult
}

File: api/app/core/schemas/twitter_downloader.py
from typing import List

from pydantic import BaseModel, constr

from app.v1.constants import RegexPatterns
from app.core.schemas.camel_case import CamelModel


class TwitterVideoDownloaderSchema(BaseModel):
    """
    Represents the schema for a Twitter video downloader request.

    Attributes:
        url (constr): The Twitter video downloader URL.
    """

    url: constr(regex=RegexPatterns.TWITTER_URL_REGEX)


class TweetMedias(CamelModel):
    """
    Represents a media object in a tweet.

    Attributes:
        format (str): The format of the media.
        width (int): The width of the media.
        height (int): The height of the media.
        resolution (str): The resolution of the media.
        url (str): The URL of the media.
        size (str): The size of the media.
    """

    format: str
    width: int
    height: int
    resolution: str
    url: str
    size: str


class TweetMetaData(CamelModel):
    """
    Represents metadata for a tweet.

    Attributes:
        tweet_url (str): The URL of the tweet.
        uploader_user_id (str): The user ID of the uploader of the tweet.
        uploader_display_name (str): The display name of the uploader of the tweet.
        upload_date (str): The date when the tweet was uploaded.
        duration (str): The duration of the tweet.
        thumbnail (str): The URL of the thumbnail of the tweet.
        title (str): The title of the tweet.
        description (str): The description of the tweet.
    """

    tweet_url: str
    uploader_user_id: str
    uploader_display_name: str
    upload_date: str
    duration: str
    thumbnail: str
    title: str
    description: str


class TwitterVideoDownloaderResult(CamelModel):
    """
    Represents the result of a Twitter video downloader.

    Attributes:
        tweet_meta_data (TweetMetaData): The metadata of the tweet.
        tweet_medias (List[TweetMedias]): The list of media objects in the tweet.
    """

    tweet_meta_data: TweetMetaData
    tweet_medias: List[TweetMedias] = []


class TwitterVideoDownloaderResponse(CamelModel):
    """
    A response model for the Twitter video downloader.
    
    Attributes:
        result (TwitterVideoDownloaderResult): The result of the video downloader.
    """

    result: TwitterVideoDownloaderResult
